### PR TITLE
feat(@angular/cli): implement deploy command

### PR DESCRIPTION
**Pull request by mgechev**
Implementation of the `ng deploy` command. Find the design document [here](https://docs.google.com/document/d/1qyVJUjoZEB3QHGBe82ogwUsnAKBv2lqWRbe8BafaGtQ/edit?usp=sharing).

// cc @vikerman @StephenFluin @clydin @alan-agius4 @filipesilva

### CODE REVIEW

File: packages/angular/cli/commands/deploy-impl.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import { ArchitectCommand, ArchitectCommandOptions } from '../models/architect-command';
+import { Arguments } from '../models/interface';
+import { Schema as DeployCommandSchema } from './deploy';
+
+export class DeployCommand extends ArchitectCommand<DeployCommandSchema> {
+  public readonly target = 'deploy';
+  public readonly error = `Cannot find "deploy" target for the specified project.
+
+You can find a package that implements deployment capabilities for your
+favorite platform on npm: https://www.npmjs.com/search?q=ng%20deploy
```
Comment (StephenFluin) [inline packages/angular/cli/commands/deploy-impl.ts]: We should provide a short list of 2-3 examples. Maybe @angular/fire, azure, etc?

Comment (mgechev) [reply packages/angular/cli/commands/deploy-impl.ts]: What would be the recommendation criteria? We don't have first party deployment builders.

**Comment by StephenFluin** [comment]
I think including the folks that we worked directly with for the initial
launch would be fair. Otherwise we could say "cloud providers of x
size/revenue/users".

File: scripts/validate-user-analytics.ts [hunk 2]
```diff
@@ -26,6 +26,7 @@ const metricsTableRe = /<!--METRICS_TABLE_BEGIN-->([\s\S]*)<!--METRICS_TABLE_END
 function _exec(command: string, args: string[], opts: { cwd?: string }, logger: logging.Logger) {
   const { status, error, stdout } = spawnSync(command, args, {
     stdio: ['ignore', 'pipe', 'inherit'],
+    maxBuffer: 1024 * 1024,
```
Comment (filipesilva) [inline scripts/validate-user-analytics.ts]: This change should be in it's own commit.

File: packages/angular_devkit/build_angular/src/deploy/index.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import { createBuilder } from "@angular-devkit/architect";
+
+const NotImplemented = `
+
+Cannot find "deploy" target for the specified project.
```
Comment (filipesilva) [inline packages/angular_devkit/build_angular/src/deploy/index.ts]: We could add this message directly on the CLI command instead of creating a dummy builder that is added to all new projects.

File: packages/angular_devkit/build_angular/builders.json [hunk 4]
```diff
@@ -16,6 +16,11 @@
       "schema": "./src/dev-server/schema.json",
       "description": "Serve a browser app."
     },
+    "deploy": {
```
Comment (filipesilva) [inline packages/angular_devkit/build_angular/builders.json]: Adding a new builder to build-angular should be a separate feature commit.

File: packages/angular_devkit/build_angular/src/deploy/index.ts (hunk 3 above)
Comment (mgechev) [reply packages/angular_devkit/build_angular/src/deploy/index.ts]: Without the architect builder the initialize phase of `ArchitectCommand` fails since it cannot find the target in `angular.json`.

Comment (filipesilva) [reply packages/angular_devkit/build_angular/src/deploy/index.ts]: Sure, but we can catch that error and show the message. It's ok to not have the default targets.

I don't quite agree with that future use of the dummy builder. The approach of using an existing builder to serve as an extension point to other builders subverts the current existing extension point which using builders themselves. With that proposed approach, a separately plugin approach would be needed just for deploy builders so that they could be used by the currently dummy builder.

File: packages/angular_devkit/build_angular/src/deploy/index.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+You should add a package that implements deployment capabilities for your
+favorite platform.
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/deploy/index.ts]: Is it intentional that this is in another new line?

File: packages/angular_devkit/build_angular/src/deploy/index.ts (hunk 3 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/deploy/index.ts]: Yeah, I agree with @filipesilva

File: packages/angular_devkit/build_angular/src/deploy/index.ts (hunk 5 above)
Comment (mgechev) [reply packages/angular_devkit/build_angular/src/deploy/index.ts]: Yep

File: packages/angular_devkit/build_angular/src/deploy/index.ts (hunk 3 above)
Comment (mgechev) [reply packages/angular_devkit/build_angular/src/deploy/index.ts]: Given the current implementation of the abstract `ArchitectCommand` it doesn't look like it's okay to not have default targets.

When someone invokes `ng deploy --help` the `initialize` method will fail because it'll not find a builder implementation.

Given the current constraints set by `ArchitectCommand.prototype.initialize`, it looks like that's the intended behavior. We can change it, sure.

Overriding the method in `DeployCommand`, catching the exception and handling it is an option but then we should reimplement part of the abstract `initialize`.

File: packages/angular_devkit/build_angular/builders.json (hunk 4 above)
Comment (mgechev) [reply packages/angular_devkit/build_angular/builders.json]: Done

File: scripts/validate-user-analytics.ts (hunk 2 above)
Comment (mgechev) [reply scripts/validate-user-analytics.ts]: Done

File: packages/angular_devkit/build_angular/src/deploy/index.ts (hunk 3 above)
Comment (filipesilva) [reply packages/angular_devkit/build_angular/src/deploy/index.ts]: Sorry I wasn't very clear. What I meant was that it is ok for users to not have (our) default targets. For instance on `--minimal` projects the `test` and `e2e` targets aren't there and that should be handled on our side.

I haven't look at how hard it is to display a custom message for a missing target. But however hard it is I will be easier than introducing new public API (the dummy builder), requiring an update (adding the dummy builder to existing angular.json). So I think we should do it in a way that doesn't require those things, whether through error catching, refactoring the command initialization, or something else. @clydin might have ideas of how to implement this.

Maybe this just means that first party Architect commands that expect the presence of targets need to have better messages in general instead of telling users their angular config is bad.

File: packages/angular/cli/models/architect-command.ts [hunk 6]
```diff
@@ -350,7 +352,8 @@ export abstract class ArchitectCommand<
         return allProjectsForTargetName;
       }
 
-      throw new Error(`Could not determine a single project for the '${targetName}' target.`);
+      throw new Error(this.missingTargetError ||
```
Comment (clydin) [inline packages/angular/cli/models/architect-command.ts]: In this case there are multiple projects with the particular target and the target is a single instance target.  The missing error shouldn't apply here as it actually found several.

File: scripts/validate-user-analytics.ts (hunk 2 above)
Comment (filipesilva) [reply scripts/validate-user-analytics.ts]: Yesterday I was looking at a similar problem in https://github.com/angular/angular-cli/pull/15176/commits/e4e04e24ac006c53f9644a4007133c080e7fd675 and now I know that `1024 * 1024` is actually the default. 

Is this change actually doing something? I guess only if `maxBuffer` was being passed as a smaller value in `opts` but I don't find `maxBuffer` referenced anywhere in scripts.

File: packages/angular/cli/commands/deploy.json [hunk 7]
```diff
@@ -0,0 +1,31 @@
+{
+  "$schema": "http://json-schema.org/schema",
+  "$id": "ng-cli://commands/deploy.json",
+  "description": "Invokes the deploy builder for a specified project. If no project is specified, the CLI will invoke the deploy builder for the default project in thw workspace.",
```
Comment (filipesilva) [inline packages/angular/cli/commands/deploy.json]: Nit: `thw` -> `the`

File: packages/angular/cli/models/architect-command.ts [hunk 8]
```diff
@@ -74,12 +75,13 @@ export abstract class ArchitectCommand<
       }
     }
 
-    if (targetProjectNames.length === 0) {
-      throw new Error(`No projects support the '${this.target}' target.`);
+    if (!options.help && (targetProjectNames.length === 0)) {
```
Comment (filipesilva) [inline packages/angular/cli/models/architect-command.ts]: Why does `!options.help` need to added to this condition and the one below it? If I call `ng deploy --help` this still looks like the error that shows up.

Comment (mgechev) [reply packages/angular/cli/models/architect-command.ts]: Since we don't have a default dummy builder, the command is going to always throw when the user invokes `ng deploy --help`.

File: scripts/validate-user-analytics.ts (hunk 2 above)
Comment (mgechev) [reply scripts/validate-user-analytics.ts]: Looks like a leftover.

**Comment by mgechev** [comment]
As discussed today, there's some inconsistent behavior in the CLI in handling the `--help` option.

If we improve it, the implementation of `ng deploy` would be simpler and more consistent with the rest of the commands. Adding it for further discussion so we can sync on the work breakdown this Thursday :).

File: scripts/validate-user-analytics.ts (hunk 2 above)
Comment (mgechev) [reply scripts/validate-user-analytics.ts]: Not a leftover. On node 10.x it's smaller so this change is required.
